src/python/pants/rules/core/list_targets.py

```diff
@@ -43,60 +48,67 @@ class List(Goal):
-        collection = await Get[HydratedTargets](Addresses, addresses)
-        if provides:
-            extractors = dict(
-                address=lambda adaptor: adaptor.address.spec,
-                artifact_id=lambda adaptor: str(adaptor.provides),
-                repo_name=lambda adaptor: adaptor.provides.repo.name,
-                repo_url=lambda adaptor: adaptor.provides.repo.url,
-                push_db_basedir=lambda adaptor: adaptor.provides.repo.push_db_basedir,
-            )
-
-            def print_provides(col_extractors, target):
-                if getattr(target.adaptor, "provides", None):
-                    return " ".join(extractor(target.adaptor) for extractor in col_extractors)
-
-            try:
-                column_extractors = [extractors[col] for col in (provides_columns.split(","))]
-            except KeyError:
-                raise Exception(
-                    "Invalid columns specified: {0}. Valid columns are: address, artifact_id, "
-                    "repo_name, repo_url, push_db_basedir.".format(provides_columns)
-                )
-
-            print_fn = lambda target: print_provides(column_extractors, target)
-        else:
-
-            def print_documented(target):
-                description = getattr(target.adaptor, "description", None)
-                if description:
-                    return "{0}\n  {1}".format(
-                        target.adaptor.address.spec, "\n  ".join(description.strip().split("\n"))
-                    )
-
-            print_fn = print_documented
-    else:
-        # Otherwise, we can use only addresses.
-        collection = addresses
-        print_fn = lambda address: address.spec
-
-    with list_options.line_oriented(console) as print_stdout:
-        if not collection.dependencies:
-            console.print_stderr("WARNING: No targets were matched in goal `{}`.".format("list"))
-
-        for item in collection:
-            result = print_fn(item)
-            if result:
-                print_stdout(result)
+async def list_targets(addresses: Addresses, options: ListOptions, console: Console) -> List:
+    if not addresses.dependencies:
+        console.print_stderr(f"WARNING: No targets were matched in goal `{options.name}`.")
+        return List(exit_code=0)
+
+    provides_enabled = options.values.provides
+    documented_enabled = options.values.documented
+    if provides_enabled and documented_enabled:
+        raise ValueError(
+            "Cannot specify both `--list-documented` and `--list-provides` at the same time. "
+            "Please choose one."
+        )
 
+    if provides_enabled:
+        targets = await Get[Targets](Addresses, addresses)
+        addresses_with_provide_artifacts = {
+            tgt.address: tgt[ProvidesField].value
+            for tgt in targets
+            if tgt.has_field(ProvidesField) and tgt[ProvidesField].value is not None
```

Comment:
Can this be `tgt.get(ProvidesField, None)`?

Reply:
This could alternatively be:

```python
if tgt.get(ProvidesField).value is not None
```

or, to be more explicit:

```python
if tgt.get(ProvidesField, default_raw_value=None).value is not None
```

I went with first filtering via `tgt.has_field` as a premature optimization to avoid having `tgt.get` call the constructor `ProvidesField(raw_value=None, address=tgt.address)`. But, I realized that's not even much of a perf win due to other costs from this setup.

I'll simplify.

--

The takeaway is that `tgt.get(ProvidesField, None)` _does not_ return `Union[ProvidesField, None]`. Rather, `tgt.get(ProvidesField, default_raw_value=None)` returns `ProvidesField` because it calls `ProvidesField(raw_value=default_raw_value)` in the missing case.